### ISSUE TITLE
Poisson Loss in Random Forest Mode 

### ISSUE THREAD

**Issue by mayer79** [dmlc/xgboost#5423]
With MSE loss, the random forest mode seems to work well. However, when switching to "count:poisson" (and also "reg:gamma") loss, the model is completely off. The distribution of the predictions is heavily biased.

```
set.seed(1)
n <- 1000
x1 <- seq_len(n)
x2 <- rnorm(n)
X <- cbind(x1, x2)
y <- rpois(n, x1 / 1000 + x2^2)

library(xgboost)
library(MetricsWeighted)

dtrain_xgb <- xgb.DMatrix(X, label = y)

#=======================================================
# WITH MSE loss
#=======================================================

# xgboost with random forest like parameters
param <- list(max_depth = 10,
              learning_rate = 1,
              objective = "reg:linear",
              subsample = 0.63,
              lambda = 0,
              alpha = 0,
              colsample_bylevel = 1/3)

fit_xgb <- xgb.train(param,
                     dtrain_xgb,
                     watchlist = list(train = dtrain_xgb),
                     nrounds = 1,
                     num_parallel_tree = 500)

pred_mse <- predict(fit_xgb, X)

summary(pred_mse)
# Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
# 0.01976  0.69959  1.06019  1.51380  1.74806 11.24593

# MSE-based R-squared
r_squared(y, pred_mse) # 0.8228559

# Poisson-based R-squared
r_squared_poisson(y, pred_mse) # 0.6900247

#=======================================================
# WITH Poisson loss
#=======================================================

# xgboost with random forest like parameters
param <- list(max_depth = 20,
              learning_rate = 1,
              objective = "count:poisson",
              subsample = 0.63,
              lambda = 0,
              alpha = 0,
             # max_delta_step = 0.7,
              colsample_bylevel = 1/3)

fit_xgb <- xgb.train(param,
                     dtrain_xgb,
                     watchlist = list(train = dtrain_xgb),
                     nrounds = 1,
                     num_parallel_tree = 500)

pred_poi <- redict(fit_xgb, X)

summary(pred_poi)
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.3325  0.6366  0.8874  0.8054  0.9949  1.0069 

# MSE-based R-squared
r_squared(y, pred_poi) # -0.02036563

# Poisson-based R-squared
r_squared_poisson(y, pred_poi) # 0.01646716
```

**Comment by trivialfis** [comment]
Maybe the `max_delta_step` has something to do with the convergence.  I'm not sure if it's like the proof  for `softmax` function.

**Comment by mayer79** [comment]
In another example with different data and parameters, the result is not too different from LightGBM's rf mode, but ironically worse than using ranger with MSE loss. Poisson deviance should be low, r_squared is the proportion of deviance explained (should be high). These values are calculated on an independent validation data set. 

#                    pred_xgb      pred_lgb pred_ranger
# deviance  1.05753011 1.04356883  1.02037717
# r_squared 0.04407729 0.05669717  0.07766058

The code is in https://github.com/mayer79/random_forest_benchmark/blob/master/r/poisson.R

**Comment by mayer79** [comment]
With slightly other data structure (but similar tree settings; same link above but new commit), again garbage (training is over very quickly).

```
#            pred_xgb   pred_lgb pred_ranger
# deviance   3.470014 1.17156879  1.13810495
# r_squared -1.788888 0.05839744  0.08529269
```

Hmm.

**Comment by RAMitchell** [comment]
In random forest mode we can only perform one Newton step to minimise the objective function. In the case of squared error a single step is sufficient. This might be the source of bias. I have thought about how to fix this, maybe by refreshing the trees in subsequent boosting steps.

**Comment by mayer79** [comment]
This might be indeed the reason why performance is not too good. But I guess lightGBM's rf mode would suffer the same issue. However, its performance is not negative in all cases I have tested. I will try to test more delta_step values to see if it is just a poor parametrization by myself.
